### ISSUE TITLE
Improve documentation of health check type "process"

### ISSUE THREAD

**Issue by vanschelven** [cloudfoundry/docs-dev-guide#487]
The current documentation says

> For a process health check, Diego ensures that any process declared for the app stays running. If the process exits, Diego stops and deletes the app instance.

I'm quite sure that I have observed "in the wild" that "if the process exits" actually should read "if the process and all its children exit", but I would like to confirm this by reading the code that implements the process health check.

1. could someone point me to this code?
2. (assuming my assumption was correct) I'm not sure how I feel about the "and all its children" part... in a typical worker setup, where one worker spawns many others, the setup has become unhealthy and the whole cell should be restarted if the main process has gone away. 
3. assuming the behavior isn't changed, at least the documentation should be brought inline.

**Comment by anita-flegg** [comment]
Thanks @vanschelven , I will take a look at the documentation. You might also want to ask these questions in the Cloud Foundry Slack channels.

**Comment by vanschelven** [comment]
> slack channels

"This team's administrator has not enabled email signups."

**Comment by anita-flegg** [comment]
@ameowlia , @gerg, would you please suggest a reviewer for this?
Thanks!
(I hate to see this good stuff languish :)

**Comment by ameowlia** [comment]
Thanks for the ping @anita-flegg , this falls under my WG.

I am not sure which ones of these is true:
* "if the process exits" 
* OR "if the process and all its children exit"

I will ping the interrupt in CFF slack and ask them to investigate.

**Comment by anita-flegg** [comment]
Thanks, @ameowlia :)

**Comment by geofffranks** [comment]
Hi @vanschelven !

I believe the docs are correct. 

> For a process health check, Diego ensures that any process declared for the app stays running. If the process exits, Diego stops and deletes the app instance.

The health check looks at the primary process that was spawned inside the app container. If it exits, diego/garden believe this to mean that the application inside it has crashed. As a result, diego instructs garden to clean up the container. Garden notifies containerd to stop all processes in the container namespace. The diego healthcheck only needs to know about the primary process in order to perform its healthcheck. 

Under normal circumstances an app process inside the container should stay up indefinitely until it is told to exit by diego/garden. When this happens, the container is cleaned up, and the healthcheck plays no role. As part of this cleanup, garden calls containerd to delete the container and signal/stop all processes inside it.

During the case of a Diego Task (rather than LRP/App container), health checks are not spawned, as this is designed to be an ephemeral container.

**Comment by vanschelven** [comment]
> The health check looks at the primary process that was spawned inside the app container. If it exits, diego/garden believe this to mean that the application inside it has crashed. As a result, diego instructs garden to clean up the container. Garden notifies containerd to stop all processes in the container namespace. The diego healthcheck only needs to know about the primary process in order to perform its healthcheck.

Could you point me to the code that implements the above? Given my observations "in the wild" I would really like to have some additional evidence that this is how it works.

**Comment by geofffranks** [comment]
Http and TCP health checks are [configured up here in the executor](https://github.com/cloudfoundry/executor/blob/main/depot/transformer/transformer.go#L669). If a workload doesn't have an http or tcp health check specified, it gets the default behavior of process monitoring.  HTTP and TCP health checks are handled as [codependent steps in executor](https://github.com/cloudfoundry/executor/blob/main/depot/transformer/transformer.go#L531-L534). They are wrapped alongside the main process in an ifrit runner that will kill all of them if any of them exit.

Process monitoring happens [slightly differently using monitors, rather than checks](https://github.com/cloudfoundry/bbs/blob/main/doc/lrps.md#monitoring-health). If specified, these monitors are also run as codependent steps in executor. If a monitor is not specified, there are no subteps for the LRP, and only the `action` of the LRP is used. When that [exits/crashes/no longer exists](https://github.com/cloudfoundry/executor/blob/main/depot/containerstore/storenode.go#L635), the [container is cleaned up](https://github.com/cloudfoundry/executor/blob/main/depot/containerstore/storenode.go#L543).

**Comment by vanschelven** [comment]
> If a monitor is not specified, there are no subteps for the LRP, and only the `action` of the LRP is used. When that [exits/crashes/no longer exists](https://github.com/cloudfoundry/executor/blob/main/depot/containerstore/storenode.go#L635),

That assumes that `process.Wait`, from `ifrit`, is implemented in the way that you mention, namely: process-death leads to `Wait` returning, rather than (as I have observed in my live apps) death of the process and all of its children. The documentation of `ifrit` is a bit too concise for me to figure it out and I'm out of my depth with `go` here...

**Comment by vanschelven** [comment]
https://github.com/tedsuo/ifrit/blob/master/process.go#L79 this is the ifrit I mentioned

**Comment by geofffranks** [comment]
> namely: process-death leads to Wait returning, rather than (as I have observed in my live apps) death of the process and all of its children. 

The app process is not run via ifrit, it is run via garden-runc calling containerd. Executor calls garden to start the job, and the ifrit job there is waiting on the result of garden's container processes for the healthcheck, or the `action` process. Once one of those fails, ifrit's notifiers trigger the other steps to start cleaning up. This tells garden to stop the container completely.

The documentation in question states:

> For a process health check, Diego ensures that any process declared for the app stays running. If the process exits, Diego stops and deletes the app instance.

If the process exits, diego stops and deletes that app instance, which includes any child processes it spawned. There is no other trigger in Diego/garden to watch child processes. However, when the main process exits, garden cleans up the container which leads to the destruction of any child processes still running as they are now orphaned and their parent is the init process for the container.

**Comment by vanschelven** [comment]
Whatever the cause may be, the following assertion is untrue, at least in my environment:

> If the process exits, diego stops and deletes that app instance, which includes any child processes it spawned. 

This can easily be reproduced:

A parent:
```
# die-after-spawn.py 
import subprocess

subprocess.Popen(["python", "live-forever.py"])

# (exit is implicit), this is the last line of the file

```

a child:
```
# live-forever.py 
from time import sleep

while True:
    sleep(1)
    print("still alive")
```

a manifest snippet

```
     health-check-type: process
     no-route: true
     command: python die-after-spawn.py
```

logs:

```
still alive
still alive
still alive
```
no crashes are reported, and `live-forever.py` (but not die-after-spawn.py) shows up in the output of `ps aux`.

**Comment by anita-flegg** [comment]
@geofffranks, would you respond to this concern, please? I would like to either make a doc update or close this issue. 
Thanks!

**Comment by geofffranks** [comment]
Thanks for the reminder @anita-flegg  - Looks like I was wrong and @vanschelven was right. We should update the docs to read "if the process and all its children exit"

**Comment by anita-flegg** [comment]
Resolved under PR #495

**Comment by anita-flegg** [comment]
Thanks, @geofffranks  -- much appreciated.
